### ISSUE TITLE
Does our design need to account for problems unique to unchecked code?

### ISSUE THREAD

**Issue by kevinb9n** [jspecify/jspecify#115]
It has come up in conversation many times that "yes, but if that dependency code isn't being checked, then &lt;thing to worry about&gt;", and I am trying to understand how much we do or don't need to worry about this concern and why.

The situation: when analyzing a source file, a tool will never know whether the file's dependencies were analyzed in the same way, different way, or not at all. If the annotations we see are coming from overlay files, it's probably especially unlikely that this actual verification happened.

Similarly, even if that dependency is checked, if it's not a final class, then it might have _subtypes_ that weren't analyzed by a suitably compliant tool.

Similarly, it's always possible that the foreign code was checked, but that warnings/errors were suppressed.

Putting it together, I think that it all boils down to this single truth: when you depend on other code, it is always possible that its implementation may fail to adhere to the promises made by its specification (annotations). This seems like a basic problem that cuts across everything we are dealing with.

But when it does fail, it seems generally possible to pinpoint blame. There is some mismatch either between an implementation and its own contract, or between a contract and a supercontract, or between a contract expressed in overlay files and the code it's trying to describe.

Put simply, when implementation clashes with spec, we know that it will cause problems, and we know that the proper remedy is for them to be brought in line with each other.

At this point I am not clear on whether the issues just described actually need to have some influence on our design choices at all. By analogy, if I receive an `Iterator<Long>` it is true that it might actually be an `Iterator<String>` (heap pollution), but it's not "my" problem if it is.

**Comment by lazaroclapp** [comment]
> I am trying to understand how much we do or don't need to worry about this concern and why.

This seems to me like a place where the answer differs between sound and unsound tools. The answer is that an _unsound_ tool needs not care - NullAway won't have an issue with code that describes itself as "annotated" but was checked to dramatically more lax standards than it itself uses, mostly because it already assumes the best case out of unannotated code - but a _sound_ tool absolutely has to care. Because a sound tool must treat unannotated code pessimistically, it follows that it would want to treat "annotated code analyzed by an unsound tool" pessimistically as well, specially if said annotated code interacts with unannotated code further down the line.

I think - I am not sure, but I think, and it might undermine the whole endeavor to an extent if this is true - that there are cases where the code conforms to the spec, the tools conform to the spec, and yet the output of an unsound tool is not useful to a sound one. 

Consider the following three classes, each checked or not by different tools:

```
// Unchecked
class A {
   public static void consume(Object o) {
      log(o.toString()); // DEREF
   }
}
```

```
// checked by NullAway
class B {
   public static void takesNonNull(Object o) {
      // Safe by NullAway's and Jspecify logic due to 
      // optimistic treatment of unknown nullness.
      A.consume(o); 
   }
}
```

```
// checked by Checker Framework
class C {
   public static void call1(Object o) {
      // CF throws ERROR, due to pessimistic assumptions on consume
      A.consume(o); 
   }
   public static void call2(Object o) {
      // CF can't see this is a problem, because it was "misled" by NullAway
      B.takesNonNull(o); 
   }
}
```

Is it NullAway's fault? I'd argue it's not. NullAway checked the code of `B` for null dereferences and guarantees there are none in checked code, but makes no promises about the code in `A` and it's very open about the fact that it will make optimistic assumptions on it.

Is it Checker Framework's fault for trusting the code in `B`? Maybe, but if it is, then CF can't rely on `jspecify`. If it isn't, then CF through no fault of its own failed to fulfill the tool's objective, which is to catch _all_ potential NPEs in the whole-program (soundness).

Heap pollution is a violation of an abstraction. But the example above is two tools with slightly different "promises" on what the absence of `@Nullable` in `B.takesNonNull`'s argument means. And it sounds to me that currently both types of promises are compatible with what jspecify says...

Maybe @wmdietl or @mernst can't correct me there, though. I'll note that this is not necessarily the only place in which CF relies on unsound input, since humans are unsound and so therefore are `.astub` files. So maybe it's not such a big problem to get annotations out of other tools that aren't quite up to their standard of soundness but "usually correct" in practice and noting that CF is sound only when it gets to be the one to analyze all the code?

**Comment by kevin1e100** [comment]
@lazaroclapp I'm a little unsure about your example but I'm guessing `B.takesNonNull` would be declared `void takesNonNull(@Nullable Object o)` and all type usages in class `C` are implicitly treated as `@NonNull` by checkerframework?

Regardless of the intended annotations: the NPE in the example appears in unchecked code, IIUC.  Fundamentally, that can always happen, even when using sound tools on other code: tools can only make promises about code they see, and only assuming other code behaves as expected.  So if not all code is actually checked with Checkerframework, for instance, then NPEs are possible, _even in checked code_.  Whether NullAway is used on code that CF didn't check doesn't change that.  Note that Kotlin is similar: the moment Java code is in the picture, NPEs become possible, _even in Kotlin code_, and even if the referenced Java code is annotated (since the annotations may be wrong).

But I do think jspecify can help a bit: what seems possible to me is that sound tools will be able to guarantee absence of NPEs _in code they check_ **assuming** `null` values never flow into checked code from unchecked code in places annotated `@NonNull` **as long as** the tool makes sure `null` values only flow out of checked code into unchecked code in places annotated `@Nullable`.  (Note that to achieve this, tools will additionally have to make conservative assumptions about values appearing in places annotated `@UnspecifiedNullness`.)

That is to say, tools will be able to give guarantees about code they check as long as annotations on other code (not checked at all, or not checked with the same tool) are "correct".  At the same time, tools can clearly identify completely unannotated code and treat it conservatively, or optimistically, as desired, and qualify their soundness guarantees by how unannotated code is treated.

One other thing to note is that jspecify annotations as discussed so far do _not_ promise absence of NPEs in annotated code: the annotations are strictly about which type usages in a class declaration allow `null` values; they say nothing about the behavior of annotated methods except what those methods should return.  Annotations discussed so far in particular don't preclude methods from explicitly throwing NPEs (in response to `null` values from `@Nullable` method parameters or otherwise), or from throwing any other exception for that matter.  (Note it may be possible to attach some sort of "I won't throw NPEs" promise to an annotation like `@DefaultNonNull`, but that's to my knowledge not been considered so far.)

**Comment by lazaroclapp** [comment]
Hi @kevin1e100 , sorry, I just re-read the example and I apparently had some things twisted around when I initially wrote it. It certainly didn't make sense as written!

I just edited it. It should be a `takesNull(@Nullable Object o)` method for `B` in order for it to work as intended. Using the edited version in my discussion below.

> Whether NullAway is used on code that CF didn't check doesn't change that.

There is one difference: CF - I believe - would treat unannotated code pessimistically, meaning it would interpret `A.consume(Object o)` as `A.consume(@NonNull Object o)`. But, NullAway would treat it as `A.consume(@Nullable Object o)` to minimize FPs. The problem is that when CF sees the NullAway annotated `B.takesNull(@Nullable Object o)`, it no longer treats the call pessimistically, because there is an explicit annotation. 

NullAway annotated code that calls `consume(...)` in a way that it's compatible to how NullAway treats `consume(...)` but incompatible with how CF would treat `consume(...)`. In a way, NullAway leaks its assumptions about `consume(...)` and forces CF to either make the same assumptions or distrust NullAway's annotations.

CF would not have an NPE when dealing will fully unannotated code, because it would treat it pessimistically (but would likely have many false positives). The moment it begins to trust NullAway/jspecify annotations, though, then (a new class of) false negatives become possible as well.

**Comment by kevin1e100** [comment]
I'm in violent agreement about this, and I'll reiterate that I see this as
WAI from my perspective.  I'll note that a slightly different example,
where the NPE occurs in `B` or even `C`, is also possible, and still WAI
from my perspective.  This all goes back to what I said about "soundness"
being relative to what annotations are present when only parts of the
program are analyzed by a given tool, which is in my understanding
consistent with how these tools operate anyway.

From my point of view, annotations can always be wrong: jspecify
annotations can also be used without any tool at all, plus there's the more
philosophical point that annotations, even when checked with a tool, can be
"wrong" (if they just happen to be consistent with checked code).  But no
tool I'm aware of that wants to give a global soundness guarantee, like,
"this entire application will never throw NPEs" will do so without having
seen all the code in the application.  That is to say, tools like CF only
trust annotations until they see the code "behind" them, so they *would* in
your example flag `A` (or possibly `B`) if they ever saw its
implementation.  While I understand the concern with that, I don't see a
way out of it without requiring all tools to apply sound reasoning
principles in implementation code, which so far we have shied away from.
Moreover, the situation is from my POV still better than if `B`'s author
uses no tool at all, since NullAway will still be able to flag some issues
in `B`'s implementation.  Plus, as alluded to previously, even if `B` was
checked conservatively, that doesn't prevent `A` from causing NPEs, just
not from dereferencing a passed-in method parameter.

Here are some other ideas: for instance, tools could "sign" any code they
have checked, and encode what kind of checking was done.  Or very simply,
NullAway users could include an annotation into their code indicating that
NullAway was used (or an attribute, like `@NullAware(checker =
"NullAway")`, and other tools could adjust their expectations if they
like.  In a way that is what we have today when folks use, e.g., CF's or
Infer's annotations, though it's certainly not always the case that the
corresponding tool was actually used.

**Comment by kevinb9n** [comment]
@lazaroclapp Any chance that you find Kevin's and my arguments compelling enough to mark this closed?

**Comment by lazaroclapp** [comment]
@kevinb9n 

I think for my part we can mark this as closed in the sense that the way jspecify is designed is clearly compatible with NullAway's objectives. Since the CF folks haven't jumped into this discussion, I don't feel qualified to advocate for the needs of the sounder side of the spectrum of tools 🙂  

I still feel the issue is a real one (which, I believe, we are in agreement about), and that it can reduce the soundness of more sound tools by laundering the optimistic assumptions about third-party code from less sound tools through the common annotations. On the other hand, I agree there is no obvious way around it.

> Here are some other ideas: for instance, tools could "sign" any code they have checked, and encode what kind of checking was done. Or very simply, NullAway users could include an annotation into their code indicating that NullAway was used (or an attribute, like `@NullAware(checker = "NullAway")`, and other tools could adjust their expectations if they like.

This might be one solution, which makes the sounder tool decision one depend on the entire dependency graph and the list of "tool attestations" for it. 

For example, if it is analyzing the following graph:

```
A [Source:CF] --> B [jar:NullAway]
             \--> C [jar:NullAway] --> D [jar:unchecked]
```

Where X:Y means "X analyzed by Y". Then the checker framework can trust NullAway on B (since it depends on no unchecked code, and we aim at "soundness module usual caveats" for first party code), but might decide to either treat C as unannotated or even fetch and re-analyze the source for C, or at least those methods that call into code in D.

Either way, this becomes arbitrarily complex, so I assume it will indeed have to be unspecified for the next few Jspecify releases?

I am fine with closing this to reduce issue clutter, unless there is interest/urgency from the CF project on figuring out a solution for this problem or introducing tooling metadata as above.

**Comment by kevinb9n** [comment]
Given that JSpecify is here to enable NON-whole-program analysis, "soundness" is by nature a local property, and soundness of an entire code path is the intersected soundness of its parts.

Suppose A depends on B, where B was not checked, but we are now checking A:

* If B's annotations are "wrong", it's not our problem; we just assume they are right.
* But If B's annotations are providing conflicting/ambiguous/nonsense information ... well, then tools might do *anything* in this situation, and I think we can assume for now that that's probably fine. We can get more specific if we feel we have to in the future.

**Comment by netdpb** [comment]
**Current decision**: The presence of JSpecify nullness annotations on an API is not a guarantee that the implementation of that API is correct with respect to nullness.

**Argument for changing**: Tools that promise whole-program sound nullness analysis will not be able to rely on annotations in libraries. But nobody is very concerned about this.

**Timing**: This must be decided before version 1.0 of the spec.

**Proposal for 1.0: Finalize the current decision.** If you agree, please add a thumbs-up emoji (👍) to this comment. If you disagree, please add a thumbs-down emoji (👎) to this comment and briefly explain your disagreement. Please only add a thumbs-down if you feel you can make a strong case why this decision will be materially worse for users or tool providers than an alternative.

**Comment by kevinb9n** [comment]
I think this decision is documented at https://github.com/jspecify/jspecify/wiki/local-analysis#local-analysis.

And also discussed in https://github.com/jspecify/jspecify/wiki/two-kinds-of-analysis#an-open-question as an "open question" (oops?).

I'm voting but it would be really really really nice if others could review the deep "principled" thoughts there to check that portrayal of how we get here.

**Comment by lazaroclapp** [comment]
"nobody is very concerned about this" is a bit of an exaggeration, I feel, given historical discussions we've had on this (including the stuff linked above). 

This decision _does_ make JSpecify less useful to sound/pessimistic tools unless they do some extra attestation about which tool was used to analyze each `@NullMarked` artifact and their particular guarantees. 

That said, from my personal (and NullAway's, I believe) point of view: "unchecked but `@NullMarked` code is trusted to be annotated correctly, and if it isn't then it should be checked at some point" seems like the only practical way to go in terms of adoption as an inter-operable standard. Voting thumbs up.
